--- ./new/server.py	2024-07-05 01:45:17.476347523 +0800
+++ ./old/server.py	2024-07-05 01:45:43.776346744 +0800
@@ -119,12 +119,10 @@
                 self.model.device,
             )
         else:
-            pass
-            #batch = self.model.batch_type.from_pb(
-            #    request.batch, self.model.tokenizer, self.model.dtype, self.model.device
-            #)
-        #max_supported_total_tokens = self.model.warmup(batch)
-        max_supported_total_tokens = None
+            batch = self.model.batch_type.from_pb(
+                request.batch, self.model.tokenizer, self.model.dtype, self.model.device
+            )
+        max_supported_total_tokens = self.model.warmup(batch)
 
         return generate_pb2.WarmupResponse(
             max_supported_total_tokens=max_supported_total_tokens
