diff --git a/spark/dl/src/main/scala/com/intel/analytics/bigdl/python/api/BigDLSerde.scala b/spark/dl/src/main/scala/com/intel/analytics/bigdl/python/api/BigDLSerde.scala
index 4ee380e35..695457b7f 100644
--- a/spark/dl/src/main/scala/com/intel/analytics/bigdl/python/api/BigDLSerde.scala
+++ b/spark/dl/src/main/scala/com/intel/analytics/bigdl/python/api/BigDLSerde.scala
@@ -97,6 +97,7 @@ private[spark] abstract class BigDLSerDeBase {
 /**
  * SerDe utility functions for BigDL.
  */
+@SerialVersionUID(727438742588L)
 object BigDLSerDe extends BigDLSerDeBase with Serializable {

   val PYSPARK_PACKAGE = "bigdl.util.common"
