--- /usr/local/lib/python3.8/dist-packages/psutil/_pslinux.py	2023-03-01 07:31:03.000000000 +0000
+++ _pslinux.py	2023-05-19 02:21:40.508582114 +0000
@@ -414,6 +414,7 @@
     mems = {}
     with open_binary('%s/meminfo' % get_procfs_path()) as f:
         for line in f:
+            line = line.replace(b':', b': ')
             fields = line.split()
             mems[fields[0]] = int(fields[1]) * 1024
 
